### DISCUSSION TITLE
Illegal dimensions for pixmap

### DISCUSSION THREAD

**Discussion by victor-ab** [pymupdf/PyMuPDF#1327] [Looking for help] [answered]
Hi!

I am getting a mupdf error for some specific documents:
```python
>>> doc[0].get_text("html")

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-17-0ab91b87cb83> in <module>
----> 1 doc[0].get_text("html")

~/miniconda3/envs/layoutlm38/lib/python3.8/site-packages/fitz/utils.py in get_text(page, option, clip, flags, textpage)
    672     tp = textpage
    673     if tp is None:
--> 674         tp = page.get_textpage(clip=clip, flags=flags)
    675     elif getattr(tp, "parent") != page:
    676         raise ValueError("not a textpage of this page")

~/miniconda3/envs/layoutlm38/lib/python3.8/site-packages/fitz/fitz.py in get_textpage(self, clip, flags)
   5604             self.set_rotation(0)
   5605         try:
-> 5606             textpage = self._get_text_page(clip, flags=flags)
   5607         finally:
   5608             if old_rotation != 0:

~/miniconda3/envs/layoutlm38/lib/python3.8/site-packages/fitz/fitz.py in _get_text_page(self, clip, flags)
   5593         self, clip: rect_like = None, flags: int = 0
   5594     ) -> "TextPage":
-> 5595         val = _fitz.Page__get_text_page(self, clip, flags)
   5596         val.thisown = True
   5597 

RuntimeError: Illegal dimensions for pixmap 0 -128
```
Same for `dict` and `raw_dic`. `text` works ok
`doc.scrub()` didn't help as well.

This is probably due to a poorly written pdf with some negative pixmap.
Do you know any way around it? Like cleaning it up in the stream.

**Comment by JorjMcKie** [comment]
can you share the file or just one problem page of it?

**Reply by victor-ab** [reply]
Can I send that to `jorj.x.mckie@outlook.de`? I can't make it public

**Reply by victor-ab** [reply]
sent

**Comment by JorjMcKie** [comment]
Sure!

Jorj

**Comment by JorjMcKie** [comment]
Looked at it.
It works fine on Windows for me, but shows the same error in Linux (under WSL).
Must be the images if choosing flags=0 (which excludes images) things work fine.

**Comment by JorjMcKie** [comment]
It also is outside PyMuPDF, but a MuPDF issue. If doing this under Linux:
```bash
$ mutool draw -o test-linux.html test-illegal-pixmap.pdf
page test-illegal-pixmap.pdf 1error: Illegal dimensions for pixmap 0 -128
warning: Ignoring error during interpretation

page test-illegal-pixmap.pdf 2
```
Whereas under Windows it does work
```bash
>mutool draw -o test-windows.html test-illegal-pixmap.pdf
page test-illegal-pixmap.pdf 1
page test-illegal-pixmap.pdf 2
```
Although some images are shown upside down etc.

**Comment by JorjMcKie** [comment]
The "image upside down" issue can at least be clarified under windows. Under Linux, again this error pops up:
```python
>>> imglist = page.get_images(True)
>>> pprint(page.get_image_bbox(imglist[0], transform=True))
(Rect(456.37799072265625, 554.771728515625, 597.3779907226562, 586.771728515625),
 Matrix(141.0, 0.0, -0.0, -32.0, 456.37799072265625, 586.771728515625))
>>> pprint(page.get_image_bbox(imglist[1], transform=True))
(Rect(445.0393981933594, 29.196847915649414, 583.93701171875, 49.10923767089844),
 Matrix(138.89759826660156, 0.0, -0.0, -19.912389755249023, 445.0393981933594, 49.10923767089844))
>>> 
```
Both transformation matrices have a negative d value and a positive a value, showing that the original images had been inserted upside down in the pdf.

**Comment by JorjMcKie** [comment, marked as answer]
Located the problem a bit better:
* using the new page method `page.get_bboxlog()` I found that there is one operation which creates an infinite bbox: namely the instruction that creates the yellow shaded area near the top.
* Why this causes a problem in Linux, but not in Windows: no idea:
```python
bboxes = page.get_bboxlog()
for i, item in enumerate(bboxes):
    print(i, item)
38 ('fill-shade', (-2147483648.0, -2147483648.0, 2147483520.0, 2147483520.0))
```
What you see in the bbox tuple, is the new definition of an infinite rectangle. If then looking at the page definition:
```python
In [11]: print(doc.xref_object(page.xref))
<<
  /Contents 28 0 R
  /MediaBox [ 0 0 612 792 ]
  /Parent 27 0 R
  /Resources <<
    /Font 1 0 R
    /ProcSet [ /PDF /Text /ImageB /ImageC /ImageI ]
    /Shading <<
      /Sh0 2 0 R
    >>
    /XObject <<
      /FormXob.093205580d257a5e510e567391a4c4de 4 0 R
      /FormXob.7c82326558f3ec6161464e23f3002a66 3 0 R
    >>
  >>
  /Rotate 0
  /Trans <<
  >>
  /TrimBox [ 0 0 612 792 ]
  /Type /Page
>>
```
exhibits a shade definition at xref 2, named "`/Sh0`".
I then more or less intuitively removed invoking this shading:
```python
>>> page.get_contents()  # get /Contents xrefs
28
>>> cont = bytearray(page.read_contents())
>>> # looking for command invoking that shading:
>>> cont.find(b"/Sh0 sh")
2751
>>> # now remove this byte string:
>>> cont[2751:2758] = b""
>>> # write back to page contents
>>> doc.update_stream(28, cont)
>>> doc.save("no-shade.pdf")
```
The resulting PDF then shows no error ... and looks almost the same 😎.

**Reply by victor-ab** [reply]
Wow, you are a PDF magician!
Thanks a lot for locating the source of the problem. 
That new `get_bboxlog` was very handy in this case, nice work there as well!

**Reply by JorjMcKie** [reply]
Uff thanks a lot for the compliment! I think I blushed.

But the serious question is what to do with the issue.
I just found that creating a page SVG image does not work either: not in Linux nor in Windows:
```python
>>> svg=page.get_svg_image()
mupdf: Overly large image
Traceback (most recent call last):
  File "<pyshell#8>", line 1, in <module>
    svg=page.get_svg_image()
  File "C:\Users\Harald\AppData\Local\Programs\Python\Python39\lib\site-packages\fitz\fitz.py", line 5666, in get_svg_image
    return _fitz.Page_get_svg_image(self, matrix, text_as_path)
RuntimeError: Overly large image
```
I think it is worth sending a bug report to MuPDF. I believe I am able to make a neutral PDF from your file that contains no confidential data ...
Let me try and send the result to you to get your ok ...

**Comment by JorjMcKie** [comment]
It worked! Here is a mutilated PDF with all text removed except for the headings. Only first page is left over.
But the error is still there:
```bash
$ mutool draw -o cleaned.svg cleaned.pdf
page cleaned.pdf 1error: Overly large image
warning: Ignoring error during interpretation

$ mutool draw -o cleaned.html cleaned.pdf
page cleaned.pdf 1error: Illegal dimensions for pixmap 0 -128
warning: Ignoring error during interpretation
```
This is how the page looks like:
![grafik](https://user-images.githubusercontent.com/8290722/137643289-ce7f85f8-a1e5-49fd-a9a7-9e8d5d8e4c66.png)

If you are ok with this, I would like to submit this file to MuPDF ...

**Comment by victor-ab** [comment]
That's OK!

Em dom., 17 de out. de 2021 às 17:12, Jorj X. McKie <
***@***.***> escreveu:

**Reply by JorjMcKie** [reply]
Thank you, ok then, I will cc you on the bug number.

**Reply by JorjMcKie** [reply]
Here is the link to the bug I entered: https://bugs.ghostscript.com/show_bug.cgi?id=704627
You can add yourself to the cc-list of that bug, and / or add your own comments there.
